packages/xo-server-load-balancer/src/plan.js

```diff
@@ -276,6 +278,189 @@ export default class Plan {
+    const allVms = filter(this._getAllRunningVms(), vm => vm.$container in idToHost)
+    const hostList = this._getVCPUHosts(allHosts, allVms)
+    const idealVcpuPerCpuRatio =
+      hostList.reduce((sum, host) => sum + host.vcpuCount, 0) / hostList.reduce((sum, host) => sum + host.cpuCount, 0)
+
+    debugVcpuBalancing('Try to apply VCPU prepositionning.')
+    debugVcpuBalancing(`VCPU count per host: ${inspect(hostList, { depth: null })}`)
+    debugVcpuBalancing(`Average vCPU per CPU : ${idealVcpuPerCpuRatio}`)
+
+    const vmsAverages = await this._getVmsAverages(allVms, idToHost)
+    const { averages: hostsAverages } = await this._getHostStatsAverages({ hosts: allHosts })
+
+    // 1. Find source host from which to migrate.
+    const sources = sortBy(
+      filter(hostList, host => (host.vcpuCount - 1) / host.cpuCount >= idealVcpuPerCpuRatio),
+      [
+        host => -host.vcpuCount / host.cpuCount,
+        // Find host with the most memory used
+        host => hostsAverages[host.id].memoryFree,
+      ]
+    )
+    debugVcpuBalancing(`Sources : ${inspect(sources, { depth: null })}`)
+
+    for (const sourceHost of sources) {
+      let deltaSource = sourceHost.vcpuCount - sourceHost.cpuCount * idealVcpuPerCpuRatio
+      if (deltaSource < 1) {
+        continue
+      }
+
+      // 2. Find destination host.
+      const destinations = sortBy(
+        filter(hostList, host => host.id !== sourceHost.id && host.vcpuCount < host.cpuCount * idealVcpuPerCpuRatio),
+        [
+          // trying to avoid migrations between pools
+          host => host.poolId === sourceHost.poolId,
+          host => host.vcpuCount / host.cpuCount,
+          host => -hostsAverages[host.id].memoryFree,
+        ]
+      )
+      debugVcpuBalancing(`Destinations : ${inspect(destinations, { depth: null })}`)
+
+      if (!destinations.length) {
+        continue // Cannot find a valid destination.
+      }
+
+      // Build VM list to migrate.
+      const sourceVms = Object.values(sourceHost.vms)
+
+      // eslint-disable-next-line no-labels
+      destinationLoop: for (const destinationHost of destinations) {
+        debugVcpuBalancing(`Host candidate: ${sourceHost.id} -> ${destinationHost.id}`)
+
+        // calculating how many vcpu source should give and how many destination should accept
+        let deltaDestination = destinationHost.vcpuCount - destinationHost.cpuCount * idealVcpuPerCpuRatio
+
+        if (
+          deltaDestination >= 0 ||
+          hostsAverages[destinationHost.id].cpu >= this._thresholds.cpu.low ||
+          hostsAverages[destinationHost.id].memoryFree <= this._thresholds.memoryFree.low
+        ) {
+          continue
+        }
+
+        // avoiding to migrate too much vCPUs for source or destination
+        let delta =
+          Math.abs(deltaSource) > Math.abs(deltaDestination)
+            ? Math.ceil(Math.abs(deltaDestination))
+            : Math.ceil(Math.abs(deltaSource))
```

Comment:
Same for ceil.

Reply:
We can imagine situations where for example one host has 19 vCPUs and nine other hosts have 10 vCPUs each. In that case, the ideal number of vCPU per CPU is 10.9, but each host with 10 vCPU would not accept to receive VMs with more than 0.9 vCPUs (which means no VM at all) if we don't round to upper int